### PR TITLE
[@types/parse] Add missing typings

### PR DESCRIPTION
**Pull request by dplewis**
Updated to reflect changes in Parse 2.2.1.

Please fill in this template.

- [X] Use a meaningful title for the pull request. Include the name of the package modified.
- [X] Test the change in your own code. (Compile and run.)
- [X] Add or edit tests to reflect the change. (Run with `npm test`.)
- [X] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [X] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [X] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

If changing an existing definition:
- [X] Provide a URL to documentation or source code which provides context for the suggested changes: https://github.com/parse-community/Parse-SDK-JS/blob/master/src/Parse.js
- [X] Increase the version number in the header if appropriate.
- [X] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

### CODE REVIEW

File: types/parse/index.d.ts [hunk 1]
```diff
@@ -961,6 +962,22 @@ subscription.on('close', () => {});
...
+    /**
+     * Enable pinning in your application.
+     * This must be called before your application can use pinning.
+     */
+    function enableLocalDatastore(): void;
+
+    /**
+     * Flag that indicates whether Local Datastore is enabled.
+     */
+    function isLocalDatastoreEnabled(): any;
```
Comment (agoldis) [inline types/parse/index.d.ts]:
```suggestion
    function isLocalDatastoreEnabled(): boolean;
```

File: types/parse/index.d.ts [hunk 2]
```diff
@@ -961,6 +962,22 @@ subscription.on('close', () => {});
      */
     function setAsyncStorage(AsyncStorage: any): void;
 
+    /**
+     * Gets all contents from Local Datastore.
+     */
+    function dumpLocalDatastore(): any;
```
Comment (agoldis) [inline types/parse/index.d.ts]:
```suggestion
    function dumpLocalDatastore(): Promise<{ [key: string]: any }>;
```

**Review by agoldis — CHANGES_REQUESTED** [review]
Please check out the suggested changes. 
Also, it would be nice to see tests, e.g.:

```typescript
async function test_local_datastore() {
    Parse.enableLocalDatastore();

    if (Parse.isLocalDatastoreEnabled()) {
        for (let key in await Parse.dumpLocalDatastore()) {
        }
    }
}
```

**Comment by agoldis** [comment]
> There is an issue I can't figure out.
> `Cannot assign to 'serverURL' because it is a read-only property. `

BTW, @dplewis when do you see the error?

**Comment by dplewis** [comment]
@agoldis Thanks for the review. I didn't realize how much was missing from here

File: types/parse/index.d.ts [hunk 3]
```diff
@@ -414,18 +437,23 @@ declare namespace Parse {
         constructor(objectClass: string);
         constructor(objectClass: new (...args: any[]) => T);
 
+        static and<U extends Object>(...args: Query<U>[]): Query<U>;
+        static fromJSON<U extends Object>(className: any, json: any): Query<U>;
```
Comment (agoldis) [inline types/parse/index.d.ts]:
```suggestion
        static fromJSON<U extends Object>(className: string, json: any): Query<U>;
```

File: types/parse/index.d.ts [hunk 4]
```diff
@@ -414,18 +437,23 @@ declare namespace Parse {
...
         addDescending(key: string): Query<T>;
         addDescending(key: string[]): Query<T>;
         ascending(key: string): Query<T>;
         ascending(key: string[]): Query<T>;
+        aggregate(pipeline: Query.AggregationOptions|Query.AggregationOptions[]): Query<T>;
+        containedBy(key: string, values: any[]): Query<T>;
         containedIn(key: string, values: any[]): Query<T>;
         contains(key: string, substring: string): Query<T>;
         containsAll(key: string, values: any[]): Query<T>;
+        containsAllStartingWith(key: string, values: any[]): Query<T>;
```
Comment (agoldis) [inline types/parse/index.d.ts]:
```suggestion
        containsAllStartingWith(key: string, values: string[]): Query<T>;
```

File: types/parse/index.d.ts [hunk 5]
```diff
@@ -454,13 +486,17 @@ declare namespace Parse {
         near(key: string, point: GeoPoint): Query<T>;
         notContainedIn(key: string, values: any[]): Query<T>;
         notEqualTo(key: string, value: any): Query<T>;
+        polygonContains(key: string, point: GeoPoint): Query<T>;
         select(...keys: string[]): Query<T>;
         skip(n: number): Query<T>;
+        sortByTextScore(): any;
```
Comment (agoldis) [inline types/parse/index.d.ts]:
```suggestion
        sortByTextScore(): this;
```

File: types/parse/index.d.ts [hunk 6]
```diff
@@ -454,13 +486,17 @@ declare namespace Parse {
...
         startsWith(key: string, prefix: string): Query<T>;
         subscribe(): LiveQuerySubscription;
+        withJSON(json: any): any;
```
Comment (agoldis) [inline types/parse/index.d.ts]:
```suggestion
        withJSON(json: any): this;
```

File: types/parse/index.d.ts [hunk 7]
```diff
@@ -621,15 +658,16 @@ subscription.on('close', () => {});
      */
     class User extends Object {
 
+        static allowCustomUserClass(isAllowed: boolean): void;
+        static become(sessionToken: string, options?: SuccessFailureOptions): Promise<User>;
```
Comment (agoldis) [inline types/parse/index.d.ts]: `SuccessFailureOptions` is incorrect here, see the list of accepted options [here](https://github.com/parse-community/Parse-SDK-JS/blob/master/src/ParseUser.js#L637)

**Review by agoldis — CHANGES_REQUESTED** [review]
@dplewis great job adding so many missing definitions! 👏🏻

I am sorry for the amount of comments, but there were quite a lot of changes. 

If you plan to submit more PRs similar to this one I'd suggest:

1. Create a separate commit for changes that just move code around without modification and mention that commit in the PR body. That will make the review easier.

2. Create a scoped by class and smaller PRs - it will increase the chance that the PR will be reviewed and approved quickly.

3. The PR had some obvious type mismatches, I suspect you're using `dts-gen` and that's fine, probably smaller set of changes would allow to validate them manually before submitting the PR for review.

**Comment by dplewis** [comment]
@agoldis I used dts-gen as a base. I really like how everything is in alphabetical order, makes it easier to find what is missing. If you want I can close this out and follow your suggestions on creating separate PR's go for it.

I be busy this week with releasing new versions of Parse. So If you could take this over that would be great

**Comment by dplewis** [comment]
@agoldis I made the requested changes and found a few more. How does this look?

**Review by agoldis — APPROVED** [review]
Thanks @dplewis!

**Pull request merged**